js/web/lib/wasm/jsep/webgpu/ops/binary-like-util.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {DataType} from '../../../wasm-common';
+import {TensorView} from '../../tensor-view';
+import {BroadcastUtil, ShapeUtil} from '../../util';
+import {GpuDataType, ProgramInfo, ProgramInfoLoader, ProgramMetadata} from '../types';
+
+import {ShaderHelper} from './common';
+
+type BuiltinFunctionName = string;
+export type BinaryCustomExpression = (expressionA: string, expressionB: string, expressionC?: string) => string;
+export type BinaryFunctionCall = BuiltinFunctionName|BinaryCustomExpression|{
+  scalar: BinaryCustomExpression;
+  vector: BinaryCustomExpression;
+};
+
+type CreateOpProgramShader =
+    (shaderHelper: ShaderHelper, inputs: readonly TensorView[], dimsOutput: readonly number[], vectorize: boolean,
+     doBroadcast: boolean, funcCall: BinaryFunctionCall, typeOutput: number, additionalImplementation?: string) =>
+        string;
+
+const createOpProgramInfo =
+    (metadata: ProgramMetadata, inputs: readonly TensorView[], funcCall: BinaryFunctionCall,
+     createOpProgramShader: CreateOpProgramShader, additionalImplementation?: string,
+     outputDataType?: number): ProgramInfo => {
+      const a = inputs.length === 3 ? inputs[1] : inputs[0];
+      const b = inputs.length === 3 ? inputs[2] : inputs[1];
+      if (outputDataType == null) {
+        outputDataType = inputs.length === 3 ? inputs[1].dataType : inputs[0].dataType;
+      }
+
+      const isBroadcast = !ShapeUtil.areEqual(a.dims, b.dims);
```

Comment:
For `where`, the output shape is the broadcasted shape of condition, X, and Y rather than the broadcasted shape of X and Y. So here you utilize the binary's logic to only calculate a & b's broadcast is not correct. For example, condition is [4, 3], X and Y both are [1], you can't get correct result using your current algorithm.

I prefer you don't change binary's logic. And implement the `where` shader standalone. It's not that complicated compared with the current change. The final `where` shader will be something like below:
```
{
...
let outputIndices = xxx; 
let condition = ${condition.getByOffset(${calculateOffset(outputIndices, condition, output)})};
let xData = ${x.getByOffset(${calculateOffset(outputIndices, x, output)})};
let yData = ${y.getByOffset(${calculateYOffset(outputIndices y, output)})};
let value = select(yData, xData, condition);
output.setByOffset(XXXX, value);
}